File: shared/objects/Bullet.js
export class Bullet {
  static get classID() {
    return 'Bullet';
  }

  static get BODY_SIZE() {
    return 40;
  }

  get CELL_SIZE() {
    return 100;
  }
  get FRICTION() {
    return 4000;
  }

  constructor() {
    this.type = 'Bullet';
  }

  update() {
    if (this.speed.length() > this.FRICTION * dt) {
      vec3.subtract(this.speed,
        this.speed
        .unit()
        .multiply(this.FRICTION * dt));
    } else {
      this.speed.init();
    }
    vec3.add(this.pos, this.speed.multiply(dt));
  }
}

File: server/native/Bullet.cc
struct Bullet: GameLevelZoneObject {
  float FRICTION;
};

void new__Bullet(const FunctionCallbackInfo<Value>& args) {
  Isolate* isolate = args.GetIsolate();
  HandleScope scope(isolate);

  Bullet *self = new Bullet;

  self->js.Reset(isolate, args[0]->ToObject());
  self->vtable = VTABLE::BULLET;
  
  Local<Object> js = Local<Object>::New(isolate, self->js);
  self->FRICTION = (float) js->GET("FRICTION")->NumberValue();

  args.GetReturnValue().Set(node_buffer_new(isolate, self));
}

void Bullet__update(Bullet *self, GameLevelZone *gameLevelZone, float dt) {
  if (self->speed.length() > self->FRICTION * dt) {
    self->speed -= self->speed.unit() * self->FRICTION * dt;
  } else {
    self->speed.init();
  }

  if (self->speed.length() < 200) {
    Local<Object> js = Local<Object>::New(isolate, self->js);
    Local<Function>::Cast(js->GET("destructor"))->Call(js, 0, nullptr);
  }

  self->pos += self->speed * dt;
}